### PR TITLE
#189 Predeployed Contracts functionality

### LINKED ISSUE #189: 🏗️ [Core Feature]: Pay EVM Gas Fees with any* tokens
We are supporting AA first and exploring using Polkadot DeFi pallets in parallel.
The goal of this issue is to add the option for projects to deploy the entrypoint contract as part of the genesis block and write docs to make it easier for projects that want to spin up a parachain that supports Account Abstraction.

## AA
* [Account Abstraction Paymaster Support](https://github.com/OpenZeppelin/polkadot-runtime-templates/issues/199)

## Polkadot DeFi Ecosystem Pallet Config
Tracked in #206  , aims to sensibly configure [HydraDX-transaction-multi-payment](https://github.com/galacticcouncil/HydraDX-node/blob/master/pallets/transaction-multi-payment/src/lib.rs).

### PR DESCRIPTION
**Pull request by KitHat**
Added functionality to add some contracts to genesis block. Currently it does not support any storage manipulations.

Fixes #189  

#### PR Checklist

- [x] Documentation

### CODE REVIEW

File: evm-template/node/src/chain_spec.rs [hunk 1]
```diff
@@ -176,7 +181,41 @@ fn testnet_genesis(
...
+        .map(|addr| {
+            (
+                addr,
+                GenesisAccount {
+                    nonce: Default::default(),
+                    balance: Default::default(),
+                    storage: Default::default(),
+                    // bytecode to revert without returning data
+                    // (PUSH1 0x00 PUSH1 0x00 REVERT)
+                    code: vec![0x60, 0x00, 0x60, 0x00, 0xFF],
```
[the remark this answers is not in the document]
Comment (KitHat) [reply evm-template/node/src/chain_spec.rs:199]: Whoops
My bad was playing around with chainspec and this is a single thing I forgot to revert.
Will fix it!

File: evm-template/node/src/contracts/abi.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    outputs: Vec<Input>,
+    #[serde(rename = "stateMutability")]
+    state_mutability: StateMutability,
+}
+
+#[derive(Deserialize)]
+#[allow(dead_code)]
+pub struct Input {
+    name: String,
+    r#type: String,
```
Comment (ozgunozerk) [inline evm-template/node/src/contracts/abi.rs:52]: neat trick, didn't know that. Thanks!

**Review by ozgunozerk — APPROVED** [review]

File: evm-template/node/src/chain_spec.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (KitHat) [reply evm-template/node/src/chain_spec.rs]: No, I had problems with adding the CLI params to the build-spec command and I was looking for the correct entrypoint.
Everything works as expected

File: evm-template/node/src/contracts/mod.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    }
+    Ok(res)
+}
+
+fn deserialize_bytecode<'de, D>(deserializer: D) -> Result<Vec<u8>, D::Error>
+where
+    D: de::Deserializer<'de>,
+{
+    let s: &str = de::Deserialize::deserialize(deserializer)?;
+    hex::decode(&s[2..]).map_err(|e| de::Error::custom(e.to_string()))
```
[the remark this answers is not in the document]
Comment (KitHat) [reply evm-template/node/src/contracts/mod.rs:101]: Yes, exactly

**Pull request merged**